### PR TITLE
Revert "nixos/gitea: requires database"

### PR DESCRIPTION
**Pull request by hrdinka**
This reverts commit ba3d0f7a95ca81be992e485b67c574b1c059c34a from https://github.com/NixOS/nixpkgs/pull/235365

This change breaks gitea for configurations where PostgreSQL is not running on the same machine. It was not necessary anyway so I opted for simply reverting this change. It is also broken on 23.05.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by hrdinka** [comment]
To be honest I am against adding `requires` dependencies at all. They implicitly place hidden dependency on often critical services without the knowledge of the user. They severely degrade the robustness of the whole system since services won’t start if the required one fails.

If I stop PostgreSQL by hand, then I want it to be stopped. With required dependencies hidden somewhere it will randomly be started again if another service gets started by a restart config, timers or socket activation. This is especially horrible if you stopped the database for critical stuff like version updates, backup/restores or for security reasons and it gets randomly started without your knowledge just in between your maintenance task. It further makes `systemctl start/stop` not behave like it should/is expected to work and it does that without any prior notification of the user. It will further teardown services that require on it, which is also not a good idea either.

Placing these kinds of dependencies between otherwise independent services is generally a very bad idea. In case someone really wants this, they can do so by adding `systemd.services.gitea.requires = [ "postgresql.service" ]` to their config. I however don’t think that it every is a good a idea to do this for the user as it degrades the robustness of the system and changes how the system works in unexpected ways.

This is also the reason why no other NixOS service does this. `requires` (and also `wants`) is only used in cases where a setup or init script is started that solely belongs to this service. When I first discovered the feature I thought it was a good idea to use it vigorously, but after getting burned many many times I can only discourage using it for anything else than noted above.

If you really want this feature, than it is best to add a `wants` dependency for yourself like mentioned above. This deals with parts of the fallout that `requires` can cause but still is not a good idea for nixpkgs.

Using the loose `after` dependency combined `wantedBy` to automatically start it with the system (just like every NixOS service does it) is the right way of doing it. If for any reason postgresql got stopped, then it should stay that way. This way the system will behave like intended. If I don’t touch it everything works like expected and if I imperatively make live changes it does what I want it to do and nothing else.

**Comment by hrdinka** [comment]
A sleep timer will just delay restarts after a crash and won’t give any guarantees about the availability of PostgreSQL. Letting it crash and restart is a valid way of handling it in rare corner cases like a system update. The service itself might correctly react to missing services like PostgreSQL, but if not a simple crash + restart is absolutely OK, we have them for a reason. The restart also fixed it for you as stated in the original post. That is what restarts/retries are for. I guess that you don’t have any problems during normal operation.

If I am altering my system declaratively I want everything to start correctly and automatically. If I am however doing manual imperative state changes, I want that every command does exactly what I expect it to do. Strong dependencies like requires break this and my system won’t behave like expected. They add unforeseen side effects to normally very clear commands and which never is a good thing. A `systemctl stop postgresql` should do exactly that. If I want more services to be stopped I will simply add them to the prompt.

With requires/wants stopped services will be started again without my knowledge or consent. Which can be disastrous especially since systemd offers a multitude of ways for activating services. Like timers, socket activation, restarts, watch dog, inotify events, dependencies to other services, …

Wants is not recommended for a more robust system. It is just more robust or better said less damaging than requires. Watchdogs, restarts and monitoring will help to achieve a more robust system without adding a ton of hidden dependencies and complexity.

**Comment by Ma27** [comment]
> How about we add a sleep timer in the ExecStartPre to make sure gitea starts after postgres is fully operational. I don't love the idea, but it's better than a hard crash.

Did you actually have an issue with that? I can't remember a single time where this happened to me (and I maintain quite a bunch of machines at work where this _could_ happen and additionally a few more personal ones). Also, `postgresql.service` is of type `notify` which means that it notifies on its own when it's ready.

That said: if I start `gitea.service` and I have a local DB I want to ensure that systemd also _starts_ `postgresql.service` (i.e. not an ordering requirement - that's what `After=` is for, but a relationship that comes into effect when starting/stopping things) and this is what Wants/Requires is for.

Now if I want to perform a maintenance on a DB server on a machine, I'd like to ensure that all services using it are stopped. That means when running `systemctl stop postgresql.service`, it also stops everything that has `Requires=postgresql.service` defined. So IMHO that choice is correct.

Now to your issue: it should be checked whether a local postgres instance is desired (e.g. by looking at the db url) and if that's not the case, skip the `requires`/`after` declaration.

But all in all :-1: to the current approach in here.

**Comment by Ma27** [comment]
I'm not OK with that. Without having read your answer in all details it's not OK to merge something when there are still open discussions, will do so later.

**Comment by happysalada** [comment]
I'm trying to come back to this with a fresh look.

- "requires" doesn't prevent a failure of gitea if postgres takes time to restart.
- the "requires" behavior is what is expected if the database is running locally. (you can't have gitea running without postgres, and so stopping gitea when postgres stops makes sense).
- The requires is problematic if the db is not running on the same machine.
- We don't have a way to prevent gitea crashing if postgres takes time to start. (adding a sleep timer to the preExecStart doesn't solve all the use cases).

Based on this I would personally add it behind an optional flag.
I'll be happy to test again on the next postgres release.
Personally, I'm happy with whatever solution we go with.

**Comment by Ma27** [comment]
So, after looking at the code it should be fine to set `requires = ` if `cfg.database.createDatabase` is `true`.

If you still have a special-case then, that's alright, but we cannot support every single case anyways. In other words, you're supposed to use `mkForce` then (that's why we have it).

> In this case wants or requisite might still be better suited. I however think that there should not be inter service dependencies like that

The feature was exactly designed for that use-case and `wants` doesn't care about the `stop` case (i.e. I want to stop my db for maintenance -> all services using this db should also be stopped).

> "requires" doesn't prevent a failure of gitea if postgres takes time to restart.

If I understand your comment correctly, then this is not the case. You had an error without `requires`.

Because `requires` ensures that the startup of a service (in conjunction with `after`) is delayed until the other service is fully started up. In the case of postgresql we have `Type = notify`. That means that postgresql can notify systemd on its own when it's ready, i.e. when it can accept new connections.

This is probably not the case for other services using `Type = simple` because these are considered "fully started up" when the ExecStart= was executed.

I exclusively referred to the case where the dependency between $service and postgresql.service is configured correctly with after/requires.

> Based on this I would personally add it behind an optional flag.

I doubt that adding yet another small flag will be useful.

As stated above in this comment, you can override the behavior without yet another option (`mkForce`) and using `cfg.database.createLocally &&  usePostgresql` rathern than `usePostgresql` as condition for requiring `postgresql.service` should be sufficient -> no need for another option.

**Comment by happysalada** [comment]
by flag, I meant add it behind a conditional, I wasn't very clear.
To be sure the change we are talking about is having the requires line be
```
      requires = optional (cfg.database.createDatabase && usePostgresql) "postgresql.service" ++ optional (cfg.database.createDatabase && useMysql) "mysql.service";
```
(happy to make a PR for that if that is the case)

very interesting about the "notify" type, thanks for clarifying!

**Comment by hrdinka** [comment]
I believe using `createDatabase` is a good enough trade-off.

> If I understand your comment correctly, then this is not the case. You had an error without requires.

If I understand it correctly the error occurred just once and was corrected by the automatic service restart. We don’t know the actual cause and I guess that the error would also have happened with `requires`. `after` does already wait until the service is fully started. This however only applies if both services are started at once (which should be the case in all automatic restarts).

Still not a fan of adding side effects to two services and I would rather let myself or the service itself choose if it wants to be stopped if the database is turned off.

@happysalada I can update this PR so you don’t have to do the extra work

Thanks you two for your time and effort!

**Comment by hrdinka** [comment]
Maybe I will give one last example of why I think `requires`/`wants` are bad for NixOS services:

Lets say you have PostgreSQL running and 10 services that depend on it. You need to stop PostgreSQl so you do:

`systemctl stop postgresql`

With how it is now in NixOS, non of the services uses require so:

* postgresql stops
* Each service will notice that postgres is gone. It can either handle the problem itself by showing a error page, having only limited functionality or alerting the admin. Or it simply crashes.
* Crashed services will continuously try to restart according to their restart config

If I know what I doing, I might have a good reason to just stop PostgreSQL and the system does not limit me in my choices and lets me simply stop it without impacting any of the other services. If I don’t know what I am doing, the worst thing that can happen is that the other services crash and restart.

However if I now do `systemctl start postgresql` PostgreSQL will be started and eventually all dependent services will restart and work as expected.

In case they use `require` all services got stopped. I only issued a `systemctl stop postgresql` command but it implicitly also stopped other services without my consent. If I now start postgres again, only it will be started. All other services will still be down and won’t start again if not told so.

This means if I don’t know what I am doing, requires will leave me with a degraded system without my knowledge. While without `requires` everything will do just how it is expected.

Yes, `systemctl stop postgresql` will notice about the other services being stopped. This can however be missed or the command could also be called in script which will also ignore the notice.

This also applies to restarts. So if I do a simple `systemctl restart postgresql` all dependent services will be stopped, but not started again (this only happens with `PartOf`).

Then there is the problem the other way around. I consider PostgreSQL as a critical service and gitea and others as lesser important services. With requires other services gain power over the state of PostgreSQL. If I deliberately turn off postgres, starting any of the other services will start it again. This especially gave me a lot of headaches in the past. Services might be started for a ton of different reasons (see my previous comments) and their start will restart postgresql while I am in the middle of an upgrade oder recovery. So not having any require/want dependencies on postgres is way safer and won’t nullify my guarantee that it will be off if I want to be.

In the end I want to have full control over my system. If everything is done automatically, the existing `wantedBy` and `after` clauses handle everything correctly. If I am doing manual work, I want every command to do the one thing it is expected to do and I don’t want any side effects that are neither foreseen nor can be turned off.

I also want to be clear, that `requires` did not fix any problem. If I understand @happysalada’s original PR correctly, then everything was handled by the automatic restart and this just happened once during a system upgrade. We don’t know why it happened but it wasn’t a real problem after all. It could be that postgres did upgrades and was not ready yet or maybe the network interfaces got reset as well during the upgrade and that was the actual problem. The assumption that `requires` waits for the service to be fully started and `after` does not, is wrong. But this assumption was the reason for this change.

**Comment by hrdinka** [comment]
>  On the account of the crash when both postgres and gitea are updated, I do think that however using requires fixes the problem. Because postgres is a service of type notify, systemd will wait until it is fully started before starting gitea.

`requires` does not help here. `after` does the same thing (when started at the same time, which is the case during an upgrade), it also waits until the service is fully started. There is no difference. `requires` is not needed for that. This assumption is not correct.

Here is the passage from the doc again:

>  … while Before= ensures that the configured unit is started before the listed unit begins starting up, After= ensures the opposite, that the listed unit is fully started up before the configured unit is started.

So `requires` did not fix your problem at all. If it happened with `after` it will happen with `requires` as well. This did not fix your problem it just added several new ones. The restart actually did take care of the problem:

> It's not a problem as gitea just restarted, but it's just a bit surprising.

So as stated previously, with just `after` everything works as expected in the manual and automatic case. Just like how you like it to be. `requires` does not add anything to this. We don’t know why it failed once for you, but `requires` does not do what you think here. You added it based on this assumption:

> From memory, requires will wait until the required service is fully started while after just makes sure the service is started after.

Which is not true. After already does that.

I guess that it failed for another reason than PostgreSQL not being ready. There are many reasons for it, it however got covered by the restart anyways.

**Comment by happysalada** [comment]
After reading the doc one more time, you are right I believe.
The requires here doesn't solve the original problem. If it happened while `After` was there, then the error should happen with `Requires` as well.

While semantically correct (gitea will not work without a functioning postgres), it can add some complexity for people doing manual stops.
(the systemd docs, mention that restart should be covered
| Besides, with or without specifying After=, this unit will be stopped (or restarted) if one of the other units is explicitly stopped (or restarted).
)

I'm happy with removing this.

Thanks again for taking the time to discuss this issue!

**Comment by Ma27** [comment]
> While semantically correct (gitea will not work without a functioning postgres), it can add some complexity for people doing manual stops.

Well I don't want my software to misbehave in weird (perhaps unexpected) ways by ignoring that the database just went offline. It is a perfectly reasonable default to make sure this doesn't happen.

> Which is not true. After already does that.

Not necessarily: `After=` only takes effect if both units are started in the same transaction (e.g. `systemctl start postgresql gitea`). So with both services being down and only starting `gitea.service` (without `Requires=` being set), `postgresql.service` will remain inactive.

> So if I do a simple systemctl restart postgresql all dependent services will be stopped, but not started again

This is not correct:

```
[root@cloud:/etc/systemd/system]# systemctl cat wiki-js|grep Requires=
Requires=postgresql.service

[root@cloud:/etc/systemd/system]# systemctl restart postgresql

[root@cloud:/etc/systemd/system]# systemctl is-active wiki-js
active
```

> PostgreSQL will be started and eventually all dependent services will restart and work as expected.

Nope. For restarts there's a rate-limiting (StartLimitIntervalSec/StartLimitBurst and Default*) and if that rate-limiting is reached, restart attempts will be given up (which is btw perfectly reasonable because then you can configure your monitoring to send an alert if a critical service has reached a failed state). In other words, you still have to restart here (unless you're in some kind of crash loop).

Right now we have explicit ordering, that means that gitea will only start up correctly if a necessary precondition is met, i.e. postgresql is there. And if this precondition "disappears", gitea will be stopped (for e.g. maintenance work).

So again, I think that this is the correct behavior and the proper fix for this bug (gitea not starting up with a remote database) is to place `requires = ` behind `createLocally && usePostgres`.

If you want to have a general change in nixpkgs about this - this is by far not the only service using requires for that purpose - please file a ticket for that discussion, but arbitrarily doing different things on different services for the same use-case isn't very nice.

**Comment by happysalada** [comment]
interesting fact about the starting of services in a transaction!

I think the case that is unexpected is `systemctl stop postgres` (gitea becomes inactive), then `systemctl start postgres` (gitea is still inactive).
This needs to be kept in mind when doing maintenance work.

**Comment by happysalada** [comment]
@hrdinka I understand this is not your ideal scenario, but do you want to modify the current PR to put the requires option behind a createLocally conditional ? It improves the status quo, I think you would agree. I'm happy to do it as well if you would rather not, no worries at all.

**Comment by hrdinka** [comment]
The PR is now updated.

@happysalada I have been sick + on the road traveling and could not respond sooner

Thanks for everyone’s time and work, have a nice day

**Comment by happysalada** [comment]
@hrdinka no worries at all, thank you for making the change.

@Ma27 im on mobile, so i cant approve through github, but the change looks good to me. Ill let you do the merging if youre comfortable with the change since you are the maintainer.

**Review by Ma27 — APPROVED** [review]
Thanks a lot, this seems good to me!

Also @hrdinka no worries! I guess most of us do this in their free time, so there's no obligation to finish things up quickly etc. :)

**Pull request merged**
